api/client/connection.go

```diff
@@ -83,21 +97,30 @@ type Response struct {
 }
 
 func (c *Connection) Do(req *Request) (*Response, error) {
-	httpreq, err := req.HTTPRequest()
-	if err != nil {
-		return nil, err
-	}
-	res, err := c.client.Do(httpreq)
-	if err != nil {
-		return nil, err
-	}
-	if res.StatusCode < 200 || res.StatusCode > 299 {
-		err = parseError(res)
+	for i := 0; ; i++ {
+		httpreq, err := req.HTTPRequest()
+		if err != nil {
+			return nil, err
+		}
+		res, err := c.client.Do(httpreq)
+		if err != nil {
+			return nil, err
+		}
+		if res.StatusCode < 200 || res.StatusCode > 299 {
+			err = parseError(res)
+			if i == 0 && res.StatusCode == 401 && err.(*ErrorResponse).Err.Error() == "invalid token" {
```

Comment:
Retrying the request like this unfortunately is not going to work if `req.Body` is an `io.Reader`. One way to handle that is with a lightweight preflight request like `client.Connection.AuthIdentity`, either for all requests or for just `io.Reader` requests. Another way to handle it is to rely on the token's expiration time to trigger the refresh. (I like the expiration time approach because it avoids the extra request.)

Reply:
Excellent point.

Reply:
If the body is the request I'm going to have it use a track/recorder. package netutil used to have something similar for similar situations. I believe the standard library does this for redirects in requests.

Reply:
That works too!

Reply:
Alright I added request body replay functionality. This will actually save us on redirects as well since we were passing non-recognized io.Readers to http request.